src/Microsoft.CodeQuality.Analyzers/Core/ApiDesignGuidelines/StringBuilderAppendShouldNotTakeSubstring.Fixer.cs

```diff
@@ -0,0 +1,303 @@
+                string title = null;
+
+                switch (fixAllContext.Scope)
+                {
+                    case FixAllScope.Document:
+                    {
+                        var diagnostics = await fixAllContext.GetDocumentDiagnosticsAsync(fixAllContext.Document)
+                            .ConfigureAwait(false);
+                        diagnosticsToFix.Add(new KeyValuePair<Project, ImmutableArray<Diagnostic>>(fixAllContext.Project, diagnostics));
+                        title = string.Format(titleFormat, "document", fixAllContext.Document.Name);
+                        break;
+                    }
+                    case FixAllScope.Project:
+                    {
+                        var project = fixAllContext.Project;
+                        var diagnostics = await fixAllContext.GetAllDiagnosticsAsync(project).ConfigureAwait(false);
+                        diagnosticsToFix.Add(new KeyValuePair<Project, ImmutableArray<Diagnostic>>(project, diagnostics));
+                        title = string.Format(titleFormat, "project", fixAllContext.Project.Name);
+                        break;
+                    }
+                    case FixAllScope.Solution:
+                    {
+                        foreach (var project in fixAllContext.Solution.Projects)
+                        {
+                            var diagnostics = await fixAllContext.GetAllDiagnosticsAsync(project).ConfigureAwait(false);
+                            diagnosticsToFix.Add(new KeyValuePair<Project, ImmutableArray<Diagnostic>>(project, diagnostics));
+                        }
+
+                        title = "Fix all occurrences in the solution";
+                        break;
+                    }
+                    case FixAllScope.Custom:
+                        return null;
+                    default:
+                        break;
+                }
+
+                return new FixAllStringBuilderAppendShouldNotTakeSubstringDocumentChangeAction(title, fixAllContext.Solution, diagnosticsToFix);
+            }
+        }
+
+        private class FixAllStringBuilderAppendShouldNotTakeSubstringDocumentChangeAction : CodeAction
+        {
+            private readonly List<KeyValuePair<Project, ImmutableArray<Diagnostic>>> _diagnosticsToFix;
+            private readonly Solution _solution;
+            public FixAllStringBuilderAppendShouldNotTakeSubstringDocumentChangeAction(string title, Solution solution,
+                List<KeyValuePair<Project, ImmutableArray<Diagnostic>>> diagnosticsToFix)
+            {
+                Title = title;
+                _solution = solution;
+                _diagnosticsToFix = diagnosticsToFix;
+            }
+
+            public override string Title { get; }
+
+            protected override async Task<Solution> GetChangedSolutionAsync(CancellationToken cancellationToken)
```

Comment:
>GetChangedSolutionAsync [](start = 52, length = 23)

Logic is too complicated here - I would instead recommend writing a simple code fix that just fixes a single instance of your diagnostic. I think it is very unlikely to have nested invocations which have this diagnostic, and even in rare cases where it does happen, I would be fine if the FixAll bails out and user has to fix such cases manually. WellKnownBatchFixer can very nicely handle cases where diagnostics are on different statements, as you will not have any merge conflicts.

Reply:
What's supported here is the pattern

```csharp
sb.Append(someString.Substring(12,3)
    .Append(someString.OtherSubstring(0, 2)
    .Append(somethingElse)
    .Append(anotherOne.Substring(3);```

This IMHO is a quite common case as it's better to append different parts one after another than to create a combined string first before.
The BatchFixer can't solve this, and running "Fix All" just ignores the whole chain, returning the unchanged input.